cpp/include/cudf/detail/iterator.cuh

```diff
@@ -313,6 +313,72 @@ struct scalar_pair_accessor : public scalar_value_accessor<Element> {
   }
 };
 
+/**
+ * @brief Utility to fetch representation data type of specified Element type.
+ *
+ * By default, rep_type_impl<Element>::type == Element.
+ */
+template <typename Element, typename Ignore = void>
+struct rep_type_impl {
+  using type = Element;
+};
+
+/**
+ * @brief Utility to fetch representation data type of specified Element type.
+ *
+ * For fixed_point types, rep_type_impl<Element>::type == Element::rep.
+ */
+template <typename Element>
+struct rep_type_impl<Element, std::enable_if_t<cudf::is_fixed_point<Element>(), void>> {
+  using type = typename Element::rep;
+};
+
+template <typename Element>
+using rep_type_t = typename rep_type_impl<Element>::type;
+
+/**
+ * @brief Pair accessor for scalar's representation value and validity.
+ *
+ * @tparam Element The type of element in the scalar.
+ */
+template <typename Element>
+struct scalar_representation_pair_accessor : public scalar_value_accessor<Element> {
+  using super_t    = scalar_value_accessor<Element>;
+  using rep_type   = rep_type_t<Element>;
+  using value_type = thrust::pair<rep_type, bool>;
+
+  scalar_representation_pair_accessor(scalar const& scalar_value)
+    : scalar_value_accessor<Element>(scalar_value)
+  {
+  }
+
+  /**
+   * @brief returns a pair with representative value and validity of the scalar.
+   *
+   * @throw `cudf::logic_error` if this function is called in host.
+   *
+   * @return a pair with representative value and validity of the scalar.
+   */
+  CUDA_DEVICE_CALLABLE
+  const value_type operator()(size_type) const
+  {
+    return {get_rep<Element>(), super_t::dscalar.is_valid()};
+  }
+
+ private:
+  template <typename E, std::enable_if_t<std::is_same<E, rep_type>::value, void>* = nullptr>
+  CUDA_DEVICE_CALLABLE E get_rep() const
+  {
+    return super_t::dscalar.value();
+  }
+
+  template <typename E, std::enable_if_t<!std::is_same<E, rep_type>::value, void>* = nullptr>
+  CUDA_DEVICE_CALLABLE rep_type get_rep() const
+  {
+    return super_t::dscalar.rep();
+  }
```

Comment:
This looks great! You could also SFINAE on the existence of `E::rep()`, but I think what you have here makes sense.

Reply:
I guess one downside is that this method assumes that all types `T` where `std::is_same_v<T, device_storage_type_t<T>>` is `false` also have a `rep()` method on their corresponding scalar type. That may or may not be true. For instance, if we ever made `device_storage_type_t<timestamp*>` to be `int64`, then this code would fail because the timestamp scalar doesn't have a `rep()` method (yet). Granted, if we made that change to `device_storage_type_t` then we should likewise add a `rep()` method to the appropriate scalars.

Reply:
>  this method assumes that all types T where std::is_same_v<T, device_storage_type_t<T>> is false also have a rep() method on their corresponding scalar type. That may or may not be true.

The other side of that coin was whether SFINAE on `E::rep()` availability assumes that `rep()` is suitable for use here. I should probably also add a `static_assert()` that `E::rep()` returns a `device_storage_type_t`.

I'm having a little trouble getting the compile-time reflection device work. I'll update here when that's sorted out.

Reply:
I have resolved this at last. 

PEBKAC. I was using `has_rep_member<decimal32>` instead of `has_rep_member<scalar_device_type<decimal32>>`.